### PR TITLE
feat: support all C data types for boolean parameter binding

### PR DESCRIPTION
**Pull request by sfc-gh-makowalski**
Previously, ReadODBC for SnowflakeBoolean only handled SQL_C_BIT
(raw u8). Per the ODBC spec, SQL_BIT should accept all numeric C
types, SQL_C_CHAR, SQL_C_WCHAR, SQL_C_NUMERIC, and SQL_C_BINARY.

Added matching on binding.value_type for all integer types (nonzero
= true), float/double (nonzero = true, non-finite values rejected),
char/wchar (parse numeric string or "true"/"false"), numeric struct
(check val != 0), and binary (rejecting buffers with lenght other than 1).

CDataType::Default is handled explicitly (maps to u8, same as Bit).

Invalid boolean values (non-finite floats, unparseable strings) return
SQLSTATE 22018 (InvalidCharacterValueForCast) via a new
InvalidBooleanValue error variant in JsonBindingError.

Includes unit tests for each conversion path through the full
make_converter -> ParamConverter -> ReadODBC + WriteJson pipeline,
including edge cases: -0.0, NaN, ±inf, multi-byte binary buffers.

SNOW-3031772

Made-with: Cursor

### CODE REVIEW

**Comment by sfc-gh-makowalski** [comment]
* **#802** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/802?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#801** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/801?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/801?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: odbc/src/conversion/boolean.rs [hunk 1]
```diff
@@ -152,12 +155,86 @@ impl WriteODBCType for SnowflakeBoolean {
...
+/// converted to a numeric value, then 0 → false, nonzero → true.
+/// Also accepts "true"/"false" literals for Snowflake compatibility.
+fn parse_str_to_bool(s: &str, c_type: CDataType) -> Result<bool, JsonBindingError> {
+    let trimmed = s.trim();
+    if let Ok(i) = trimmed.parse::<i64>() {
+        return Ok(i != 0);
+    }
+    if let Ok(f) = trimmed.parse::<f64>() {
+        return match f.is_finite() {
+            true => Ok(f != 0.0),
```
Comment (sfc-gh-pbulawa) [inline odbc/src/conversion/boolean.rs:170]: What about `-0.0`?

File: odbc/src/conversion/boolean.rs [hunk 2]
```diff
@@ -152,12 +155,86 @@ impl WriteODBCType for SnowflakeBoolean {
...
+            CDataType::UBigInt => Ok(read_unaligned::<u64>(binding) != 0),
+            CDataType::Float => {
+                let v = read_unaligned::<f32>(binding);
+                if !v.is_finite() {
+                    return UnsupportedCDataTypeSnafu {
+                        c_type: binding.value_type,
+                    }
+                    .fail();
+                }
+                Ok(v != 0.0)
```
Comment (sfc-gh-pbulawa) [inline odbc/src/conversion/boolean.rs:211]: What about `-0.0`?

File: odbc/src/conversion/boolean.rs [hunk 3]
```diff
@@ -152,12 +155,86 @@ impl WriteODBCType for SnowflakeBoolean {
...
+            }
+            CDataType::Double => {
+                let v = read_unaligned::<f64>(binding);
+                if !v.is_finite() {
+                    return UnsupportedCDataTypeSnafu {
+                        c_type: binding.value_type,
+                    }
+                    .fail();
+                }
+                Ok(v != 0.0)
```
Comment (sfc-gh-pbulawa) [inline odbc/src/conversion/boolean.rs:221]: Again, `-0.0` is missing

File: odbc/src/conversion/boolean.rs [hunk 4]
```diff
@@ -152,12 +155,86 @@ impl WriteODBCType for SnowflakeBoolean {
...
+/// Also accepts "true"/"false" literals for Snowflake compatibility.
+fn parse_str_to_bool(s: &str, c_type: CDataType) -> Result<bool, JsonBindingError> {
+    let trimmed = s.trim();
+    if let Ok(i) = trimmed.parse::<i64>() {
+        return Ok(i != 0);
+    }
+    if let Ok(f) = trimmed.parse::<f64>() {
+        return match f.is_finite() {
+            true => Ok(f != 0.0),
+            false => UnsupportedCDataTypeSnafu { c_type }.fail(),
```
Comment (sfc-gh-pbulawa) [inline odbc/src/conversion/boolean.rs]: The datatype is supported, but the value is wrong. Can you introduce a new error, e.g. `InvalidValue/InvalidBooleanValue`?

File: odbc/src/conversion/boolean.rs [hunk 5]
```diff
@@ -152,12 +155,86 @@ impl WriteODBCType for SnowflakeBoolean {
...
+            CDataType::Numeric => {
+                let n = read_unaligned::<sql::Numeric>(binding);
+                Ok(u128::from_le_bytes(n.val) != 0)
+            }
+            CDataType::Binary => {
+                let len = buffer_data_len(binding);
+                if len == 0 {
+                    return Ok(false);
+                }
+                let first = unsafe { *(binding.parameter_value_ptr as *const u8) };
```
Comment (sfc-gh-pbulawa) [inline odbc/src/conversion/boolean.rs]: The whole buffer should be checked, as any non-zero value should produce `true`

**Comment by sfc-gh-pbulawa** [comment]
Add E2E tests for the changes.

**Review by sfc-gh-pbulawa — CHANGES_REQUESTED** [review]

File: odbc/src/conversion/boolean.rs (hunk 5 above)
Comment (sfc-gh-makowalski) [reply odbc/src/conversion/boolean.rs]: Done — now rejects len != 1 with SQLSTATE 22003 (NumericValueOutOfRange).

The ODBC spec for SQL_C_BINARY → SQL_BIT says the byte length must equal the SQL data length (1 byte), otherwise SQLSTATE 22003.

Comment (sfc-gh-pbulawa) [reply odbc/src/conversion/boolean.rs]: If that's the case, let's add a length check and make it on-par with the ODBC API and the old driver. Is the old driver spec-compliant with the explicit check?

**Comment by sfc-gh-makowalski** [comment]
> Add E2E tests for the changes.

​That's added in the next PR (802), here we have unit tests only.

File: odbc/src/conversion/boolean.rs (hunk 5 above)
Comment (sfc-gh-makowalski) [reply odbc/src/conversion/boolean.rs]: Confirmed on CI: the old ODBC driver also rejects multi-byte binary buffers for SQL_BIT. Our implementation now matches both the spec and the old driver.

**Review by sfc-gh-pbulawa — APPROVED** [review]
